### PR TITLE
Add model building support for SProc mapping

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Part of #245

### CODE REVIEW

File: src/EFCore/ModelBuilder.cs [hunk 1]
```diff
@@ -88,6 +88,18 @@ public ModelBuilder()
...
     }
 
+    /// <summary>
+    ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+    ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+    ///     any release. You should only use it directly in your code with extreme caution and knowing that
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </summary>
+    [EntityFrameworkInternal]
+    public ModelBuilder(IMutableModel model)
```
Comment (ajcvickers) [inline src/EFCore/ModelBuilder.cs:98]: Why is this needed now when it was not before?

File: src/EFCore.Relational/Extensions/RelationalEntityTypeBuilderExtensions.UseSproc.cs [hunk 2]
```diff
@@ -0,0 +1,970 @@
...
+/// <summary>
+///     Relational database specific extension methods for <see cref="EntityTypeBuilder" />.
+/// </summary>
+/// <remarks>
+///     See <see href="https://aka.ms/efcore-docs-modeling">Modeling entity types and relationships</see> for more information and examples.
+/// </remarks>
+public static partial class RelationalEntityTypeBuilderExtensions
+{
+    /// <summary>
+    ///     Configures the stored procedure that the entity type would use for updates when targeting a relational database.
```
Comment (ajcvickers) [inline src/EFCore.Relational/Extensions/RelationalEntityTypeBuilderExtensions.UseSproc.cs:17]: Is "would" the correct word to use here? @roji? @bricelam?

**Review by ajcvickers — APPROVED** [review]

File: src/EFCore/ModelBuilder.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore/ModelBuilder.cs:98]: For TPH we need to create a `PropertyBuilder` for the properties on derived entity types. This requires an `EntityTypeBuilder` for the derived entity type and the proper way of creating it is through `ModelBuilder`.
We might end up making this public with https://github.com/dotnet/efcore/issues/15258

**Comment by AndriySvyryd** [comment]
Further comments can be made on https://github.com/dotnet/efcore/pull/28502

File: src/EFCore.Relational/Extensions/RelationalEntityTypeBuilderExtensions.UseSproc.cs (hunk 2 above)
Comment (bricelam) [reply src/EFCore.Relational/Extensions/RelationalEntityTypeBuilderExtensions.UseSproc.cs:17]: I'd use *will*.

*would* implies that something will still be preventing EF from actually using it.

Comment (roji) [reply src/EFCore.Relational/Extensions/RelationalEntityTypeBuilderExtensions.UseSproc.cs:17]: Why not just "uses"? It's generally simpler/better to avoid future/conditionals when we can just use the present tense.
